File: tests/differential/output/visibility.R
# What auto-prints at the top level — R's visibility rules. Run with both
# engines and compare the whole output: every line below either prints
# (the comment says what) or prints nothing.
f <- function(v) print(v)
f(1)                                   # once, not twice
g <- function() invisible(7)
g()                                    # nothing
h <- function() { x <- 3 }
h()                                    # nothing: ends in an assignment
(x <- 5)                               # 5: parentheses make it visible
x <- 6                                 # nothing
k <- function() 42
k()                                    # 42
if (FALSE) 1                           # nothing
if (TRUE) 2                            # 2
for (i in 1:2) i                       # nothing
y <- c(1, 2); y                        # 1 2
invisible(3)                           # nothing
print(4)                               # 4, once
m <- function() { print("a"); 10 }
m()                                    # "a" then 10
n <- function() { 10; invisible(11) }
n()                                    # nothing
tryCatch(invisible(1), error = function(e) 0)   # nothing
tryCatch(8, error = function(e) 0)              # 8
switch("a", a = invisible(12), b = 13) # nothing
switch("b", a = 12, b = 13)            # 13
local({ z <- 1 })                      # nothing
local({ 14 })                          # 14
p <- function(v) return(invisible(v))
p(15)                                  # nothing
p2 <- function(v) return(v)
p2(16)                                 # 16
q2 <- function(v) cat("hi\n")
q2(1)                                  # hi
1 + invisible(2)                       # 3: arithmetic is visible
w <- 0; while (w < 2) w <- w + 1       # nothing
sapply(c(1, 2, 3), function(i) invisible(i))  # 1 2 3: sapply's value is visible

File: crates/r2-engine/tests/visibility.rs
//! R's visibility rule, as `Engine::visible` after each top-level statement
//! — the flag both consoles read to decide what auto-prints. The whole-
//! output comparison with R is tests/differential/output/visibility.R.

use r2_engine::Engine;
use r2_parser::Parser;

/// `(statement, visible after it)` pairs, run in order in one engine.
fn check(cases: &[(&str, bool)]) {
    let mut e = Engine::new();
    for (src, want) in cases {
        for ex in Parser::parse(src).expect("parse ok") {
            e.eval(&ex).unwrap_or_else(|err| panic!("{src}: {}", err.msg));
        }
        assert_eq!(e.visible, *want, "after `{src}`");
    }
}

#[test]
fn values_are_visible_and_assignments_are_not() {
    check(&[
        ("x <- 5", false), ("x", true), ("(x <- 6)", true), ("x = 7", false),
        ("1 + 2", true), ("c(1, 2)", true), ("function(a) a", true),
    ]);
}

#[test]
fn loops_and_a_false_if_without_else_are_invisible() {
    check(&[
        ("for (i in 1:3) i", false), ("w <- 0; while (w < 2) w <- w + 1", false),
        ("if (FALSE) 1", false), ("if (TRUE) 2", true), ("if (FALSE) 1 else 3", true),
    ]);
}

#[test]
fn invisible_and_print_are_invisible_and_calls_pass_it_on() {
    check(&[
        ("invisible(1)", false), ("print(1)", false), ("1 + invisible(2)", true),
        // a user function returns the visibility of its last expression
        ("f <- function(v) print(v)", false), ("f(1)", false),
        ("g <- function() invisible(7)", false), ("g()", false),
        ("h <- function() { y <- 3 }", false), ("h()", false),
        ("k <- function() 42", false), ("k()", true),
        ("p <- function(v) return(invisible(v))", false), ("p(1)", false),
        ("m <- function() { print('a'); 10 }", false), ("m()", true),
    ]);
}

#[test]
fn builtins_that_run_user_code_pass_its_visibility_through() {
    check(&[
        ("tryCatch(invisible(1), error = function(e) 0)", false),
        ("tryCatch(8, error = function(e) 0)", true),
        ("switch('a', a = invisible(1), b = 2)", false),
        ("switch('b', a = 1, b = 2)", true),
        ("local({ z <- 1 })", false), ("local({ 14 })", true),
    ]);
}
